miner/collator.go

```diff
@@ -0,0 +1,312 @@
+	}
+
+	collator, collatorAPI, err := pluginConstructor(&config)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return collator, collatorAPI, nil
+}
+
+const (
+	interruptNotHandled int32 = 0
+	interruptIsHandled  int32 = 1
+)
+
+type blockState struct {
+	worker     *worker
+	env        *environment
+	start      time.Time
+	logs       []*types.Log
+	shouldSeal bool
+	snapshots  []int
+	committed  bool
+
+	// shared values between multiple copies of a blockState
+
+	interrupt *int32
+	// mutex to make sure only one blockState is calling commit at a given time
+	commitMu *sync.Mutex
+	// this makes sure multiple copies of a blockState can only trigger
+	// adjustment of the recommit interval once
+	interruptHandled *int32
+	// prevents calls to worker.commit (with a given blockState) after
+	// CollateBlock call on that blockState returns. examined in commit
+	// when commitMu is held.  modified right after CollateBlock returns
+	done *bool
+	// calling Commit() copies the value of env to this value
+	// and forwards it to the sealer via worker.commit() if shouldSeal is true
+	resultEnv *environment
+}
+
+type interruptContext struct {
+	interrupt *int32
+}
+
+func (ctx *interruptContext) InterruptState() int {
+	if ctx.interrupt == nil {
+		return InterruptNone
+	}
+
+	switch atomic.LoadInt32(ctx.interrupt) {
+	case commitInterruptResubmit:
+		return InterruptResubmit
+	case commitInterruptNewHead:
+		return InterruptNewHead
+	default:
+		return InterruptNone
+	}
+}
+
+func (bs *blockState) Etherbase() common.Address {
+	return bs.env.etherbase
+}
+
+func (bs *blockState) Header() *types.Header {
+	return types.CopyHeader(bs.env.header)
+}
+
+func (bs *blockState) AddTransaction(tx *types.Transaction) (error, *types.Receipt) {
+	if bs.interrupt != nil && atomic.LoadInt32(bs.interrupt) != commitInterruptNone {
+		if atomic.CompareAndSwapInt32(bs.interruptHandled, interruptNotHandled, interruptIsHandled) && atomic.LoadInt32(bs.interrupt) == commitInterruptResubmit {
+			// Notify resubmit loop to increase resubmitting interval due to too frequent commits.
+			// TODO figure out a better heuristic for the adjust ratio here
+			// the gasRemaining/gasLimit is not a good proxy for all collators
+			// where the
+			gasLimit := bs.env.header.GasLimit
+			ratio := float64(gasLimit-bs.env.gasPool.Gas()) / float64(gasLimit)
+			if ratio < 0.1 {
+				ratio = 0.1
+			}
+			bs.worker.resubmitAdjustCh <- &intervalAdjust{
+				ratio: ratio,
+				inc:   true,
+			}
+		}
+		return ErrInterrupt, nil
+	}
+
+	if bs.env.gasPool.Gas() < params.TxGas {
+		return ErrGasLimitReached, nil
+	}
+
+	// Check whether the tx is replay protected. If we're not in the EIP155 hf
+	// phase, start ignoring the sender until we do.
+	if tx.Protected() && !bs.worker.chainConfig.IsEIP155(bs.env.header.Number) {
+		return ErrTxTypeNotSupported, nil
+	}
+
+	// TODO can this error also be returned by commitTransaction below?
+	_, err := tx.EffectiveGasTip(bs.env.header.BaseFee)
```

Comment:
It will be checked by `commitTransaction`, isn't it? https://github.com/ethereum/go-ethereum/blob/master/core/state_transition.go#L250